README.md
# [Toxic Comment Classification Challenge](https://www.kaggle.com/c/jigsaw-toxic-comment-classification-challenge)

Paragraphs in quote boxes are from the Kaggle description.

>The [Conversation AI](https://conversationai.github.io/) team, a research initiative founded by Jigsaw and Google (both a part of Alphabet) are working on tools to help improve online conversation. One area of focus is the study of negative online behaviors, like toxic comments (i.e. comments that are rude, disrespectful or otherwise likely to make someone leave a discussion). So far they’ve built a range of publicly available models served through the [Perspective API](https://perspectiveapi.com/), including toxicity. But the current models still make errors, and they don’t allow users to select which types of toxicity they’re interested in finding (e.g. some platforms may be fine with profanity, but not with other types of toxic content).

>In this competition, you’re challenged to build a multi-headed model that’s capable of detecting different types of of toxicity like threats, obscenity, insults, and identity-based hate better than Perspective’s [current models](https://github.com/conversationai/unintended-ml-bias-analysis). You’ll be using a dataset of comments from Wikipedia’s talk page edits. Improvements to the current model will hopefully help online discussion become more productive and respectful.

## Evaluation

>Submissions are evaluated on the mean column-wise log loss. In other words, the score is the average of the log loss of each predicted column. 

Log loss is a way of measuring error when predicting a binary variable that takes into account how certain you are of the prediction. For example, for a given record where the target is 1, a prediction of 0.9 would be scored much more highly than a prediction of 0.7, because the higher score is much more certain of the correct prediction. However, a score of 0.1 would be scored much lower than a score of 0.5 for the same target of 1, because 0.1 is very certain of the *wrong* answer.

## Timeline

>March 13, 2018 - Entry deadline. You must accept the competition rules before this date in order to compete.

>March 13, 2018 - Team Merger deadline. This is the last day participants may join or merge teams.

>March 20, 2018 - Final submission deadline.

>All deadlines are at 11:59 PM UTC on the corresponding day unless otherwise noted. The competition organizers reserve the right to update the contest timeline if they deem it necessary.

## Useful Links

**Natural Language Toolkit**: NLTK is a leading platform for building Python programs to work with human language data. It provides easy-to-use interfaces to over 50 corpora and lexical resources such as WordNet, along with a suite of text processing libraries for classification, tokenization, stemming, tagging, parsing, and semantic reasoning, wrappers for industrial-strength NLP libraries, and an active discussion forum.

http://www.nltk.org

**scikit-learn Feature Extraction**: Text Analysis is a major application field for machine learning algorithms. However the raw data, a sequence of symbols cannot be fed directly to the algorithms themselves as most of them expect numerical feature vectors with a fixed size rather than the raw text documents with variable length.

http://scikit-learn.org/stable/modules/feature_extraction.html#text-feature-extraction

**Tf-idf**: In information retrieval, tf–idf or TFIDF, short for term frequency–inverse document frequency, is a numerical statistic that is intended to reflect how important a word is to a document in a collection or corpus.

https://en.wikipedia.org/wiki/Tf–idf

**Stop words**: In computing, stop words are words which are filtered out before or after processing of natural language data (text).

https://en.wikipedia.org/wiki/Stop_words




history.md

# History

## Retrospective

**Went well:**
* Fairly consistent weekly group meetings
* Steven's neural networks modeling
* Matthew's packaging/env/cloud/github infrastructure
* EDA notebooks
* [Final score](https://www.kaggle.com/c/jigsaw-toxic-comment-classification-challenge/leaderboard): .9746
	* Percentile: 30%

**Could be improved:**
* Get Kim and Baris up & running faster
* Focus meetings on working sessions
* Set clearer goals for each meeting
* Mid-competition change to dataset, eval metric and timeline
* Got hung up on EDA and specifically on PCA, because I wanted to be able to use the best / most comprehensive t-d matrix.
	* Should have started with a restricted analysis and expanded from there.
	* Better approach for big data -- start restricted and build up.
* More engagement in peer review on GitHub -- don't merge until we've reviewed as a group

**Thoughts for the next project:**
* Define learning objectives for each individual at beginning of project
	* Define and assign responsibilities for each project?
* Continue meeting every week -- working sessions with clear roles & goals
	* Review pull requests
* Set a deadline for EDA materials (broad split: EDA and modeling)
* Environment: hand-curate requirements.txt/environment.env
* GitHub: use pull requests as forum for reviewing, discussing, and revising work
	* I set up some GitHub integrations in slack to sub us to PRs and remind us of them daily

## January 28: Week 3 Review

Steven gave an overview of NLP and model-selection concepts, including the term-document matrix, vectorization, stemming, and parameter tuning.

Matthew, Kimberly, and Baris tried unsuccessfully to get the conda environment set up. They also ate wonderful Turkish food.

## January 21: Week 2 Review

No meeting this week because we all have lives and there is plenty of time until the competition deadlines.

**Key Results:**
* Toxic comments tend to be shorter than non-toxic comments, on average. However, there is a significant subset of toxic comments that are extremely long.
* `toxic`, `severe_toxic`, and `identity_hate` tend to be the longer toxicities.
* There seem to be 3 main types of comments, based on length:

|Length|Topic|Top Words| % of Training Set |
|--|--|--|--|
| < 250 chars| Talk | talk |
| 250-3000 chars | Article/Wikipedia Policy (?)| article, wikipedia |
| > 3000 chars| Toxicity | fuck |
![](https://files.slack.com/files-pri/T87KDTP3Q-F8UNWE7NZ/image.png)
* Toxic comments tend to have lower polarity (?) and sentiment than non-toxic ones.
* We have now fit 3 models:

|Modeling Strategy							|	Public Leaderboard Score	|
|-----|-----|
|CountVectorizer -> LogisticRegression		|	.084						|
TfidfVectorizer -> LogisticRegression 		|	.057						|
Stemming -> TfidfVectorizer -> LogisticRegression |.055 					|

## January 14: Week 1 Review

Baris and Matthew met over Google Hangouts to review progress the team has made over the past week since our initial meeting on January 7. Most of the time was spent going over Steven and Matthew's Jupyter notebooks from the past week.

**Key Results:**
* We have 96,000 records in the training set and 227,000 in the test set, consisting of raw Wikipedia comments and ratings by human workers for each comment's toxicity. Toxicity comes in the following six categories: `toxic`, `severe_toxic`, `identity_hate`, `threat`, `obscene`, and `insult`.
* The most common words in comments generally are 'article', 'page', and 'wikipedia'.
* About 10% of comments are toxic in any form, mostly `toxic`, `obscene`, and `insult`. These three classes are most common and tend to occur together.
* The most common words for `toxic`, `severe_toxic`, `obscene`, and `insult` are all fairly similar, including 'fuck' 'suck', 'nigger', 'ass', and 'shit'. The top words for `threat` and `identity_hate` are more distinct.
* The data badly needs to be cleaned. In addition to excessive punctuation, misspellings, and escape sequences, we also have some comments with automatically generated messages and information (e.g., a notice that the message was edited or removed), and outliers, such as comments that are extremely long and spammy.
* Two baseline models have been fit, both separately-fitted logistic regressions. With counts, this achieves a logloss on submission to Kaggle of 0.084, and with Tfidf vectors, 0.057. This is a good baseline on which to improve.


toxic/common.py
import wordcloud
import matplotlib.pyplot
import seaborn
from google.cloud import storage
from io import StringIO
import pandas
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

def gcs_to_df(project='mlmasters', bucket='toxic-comments', blob='sample_submission.csv'):
    """
    Gets data from Google Cloud Storage.
    """
    client = storage.Client(project=project)
    bucket = client.get_bucket(bucket)
    s = bucket.get_blob(blob).download_as_string()
    f = StringIO(s.decode())
    df = pandas.read_csv(f)
    return df

def plot_wordcloud_to_axis(cloud_string=None, title=None, ax=None, cloud_obj=wordcloud.WordCloud(random_state=42)):
    """
    Plots a wordcloud to the passed axis using the passed string.
    """
    if cloud_string:
        cloud = cloud_obj.generate(cloud_string)
        ax.imshow(cloud)
        if title:
            ax.set_title(title)
        ax.axis("off")
        return ax

def density_plot(data=None, variable=None, color_mask=None):
    """
    somewhat ggplot-like simple plotting 
    interface for densities segmented by a mask
    """
    fig = pyplot.figure(1, figsize=(10,5))
    
    ax1 = pyplot.subplot(121)
    barvals = data.groupby(color_mask).mean()[variable]
    seaborn.barplot(x=barvals.index, y=barvals, ax=ax1)
    
    ax2 = pyplot.subplot(122)
    seaborn.kdeplot(train.loc[~color_mask, variable], shade=True, color="b", label='normal', ax=ax2)
    seaborn.kdeplot(train.loc[color_mask, variable], shade=True, color="g", label='toxic', ax=ax2)
    return fig

default_params = dict(
    ngram_range=(1,1),
    stop_words='english',
    max_features=1000,
    binary=False
)

def vectorize_text(
    corpus=None, 
    output='sparse',
    vect='count',
    param_dict=default_params
):
    """
    Helper function to vectorize raw comments.
    Fits and transforms either a CountVectorizer ('count') or
    TfidfVectorizer ('tfidf') to a corpus of comments and returns
    a DataFrame with appropriately-labeled axes.    
    """
    if not isinstance(corpus, pandas.Series):
        raise ValueError('Enter a Pandas Series object for corpus')
    
    if vect == 'count':
        vectorizer = CountVectorizer(**param_dict)
    elif vect == 'tfidf':
        vectorizer = TfidfVectorizer(**param_dict)
    else:
        raise ValueError('Unrecognized vectorizer')
    
    # Fit-transform vectorizer and put into DataFrame
    print('Fitting data...')
    vectorizer.fit(corpus)
    
    print('Transforming data...')
    sparse = vectorizer.transform(corpus)
    if output.lower() == 'sparse':
        result = sparse
    elif output.lower() == 'array': 
        result = sparse.toarray()
    elif output.lower() == 'df':
        result = pandas.DataFrame(
            sparse.toarray(), 
            columns=vectorizer.get_feature_names(), 
            index=corpus.index)

        # Sort DF columns
        print('Sorting df cols...')
        word_frequencies = result.sum(axis=0)
        keep_cols = sorted(
            vectorizer.vocabulary_.copy(), 
            key=lambda k: word_frequencies[k], 
            reverse=True)
        result = result[keep_cols]
    
    print('Done')
    return result

toxic/common/data.py
"""
data
Functions for getting and preprocessing data.

Matthew Chatham
January 26
"""
import pandas as _pandas

def get_kaggle_data(file='train', verbose=False):
    """
    Get Kaggle data.
    
    Parameters
    ----------
    file : str, 'train|test' to get the corresponding data (default 'train')
    verbose : bool, True to see progress messages from BigQuery (default False)
    
    Returns
    ----------
    res : a pandas dataframe holding the train/test data
    """
    if file not in ['train', 'test']: raise ValueError('file must be train|test')
    
    try:
        res = _pandas.read_csv('{}.csv'.format(file), encoding='utf-8', index_col='id')
    except:
        query = """
        #standardSQL
        SELECT *
        FROM `mlmasters-191705.toxic_comments.{}`
        """.format(file)
        res = _pandas.read_gbq(
            query, project_id='mlmasters-191705',
            dialect='standard',
            verbose=verbose
        ).set_index('id')
        res.to_csv('{}.csv'.format(file), encoding='utf-8', index=True)
    
    return res


toxic/common/plotting.py
"""
toxic.common.plotting

Functions for plotting.

Attributes
----------
plot_wordcloud_to_axis : function, plots a wordcloud to a passed axis
density_plot : function, ggplot-like interface for plotting densities segmented by a mask
"""

import seaborn as _seaborn
import wordcloud as _wordcloud
import matplotlib as _matplotlib

def plot_wordcloud_to_axis(cloud_string=None, title=None, ax=None, cloud_obj=_wordcloud.WordCloud(random_state=42)):
    """
    Plots a wordcloud to the passed axis using the passed string.
    """
    if cloud_string:
        cloud = cloud_obj.generate(cloud_string)
        ax.imshow(cloud)
        if title:
            ax.set_title(title)
        ax.axis("off")
        return ax
    
def density_plot(data=None, variable=None, color_mask=None):
    """
    somewhat ggplot-like simple plotting 
    interface for densities segmented by a mask
    """
    fig = _matplotlib.pyplo.figure(1, figsize=(10,5))
    
    ax1 = _matplotlib.pyplo.subplot(121)
    barvals = data.groupby(color_mask).mean()[variable]
    _seaborn.barplot(x=barvals.index, y=barvals, ax=ax1)
    
    ax2 = _matplotlib.pyplot.subplot(122)
    _seaborn.kdeplot(data.loc[~color_mask, variable], shade=True, color="b", label='normal', ax=ax2)
    _seaborn.kdeplot(data.loc[color_mask, variable], shade=True, color="g", label='toxic', ax=ax2)
    return fig

toxic/common/preprocessing.py
"""
toxic.common.preprocessing

Classes and functions for processing text data.

Attributes
----------
Stemmer : class, method `stem` can be used as an analyzer for a vectorizer
vectorize_text: function, takes a series of comments and transforms to a labeled dataframe
"""

import nltk as _nltk
import pandas as _pandas
import sklearn as _sklearn

class Stemmer():
    """
    Word stemmer to use in a CountVectorizer or TfidfVectorizer.
    """
    def __init__(self, stemming=None):
        """
        stemming : str, the type of stemming algorithm to use, one of snoball|lancaster|wordnet
        """
        
        self.stemming = stemming
        if stemming == 'snowball':
            self.stemmer = _nltk.stem.snowball.EnglishStemmer()
        elif stemming == 'lancaster':
            self.stemmer = _nltk.stem.lancaster.LancasterStemmer()
        elif stemming == 'wordnet':
            self.stemmer = _nltk.stem.wordnetWordNetLemmatizer()
        else:
            raise ValueError('Unrecognized stemmer')

    def stem(self, word=None):
        """stem an individual word string"""
        
        res = ''
        if self.stemming in ['snowball','lancaster']:
            res = self.stemmer.stem(word)
        elif self.stemming == 'wordnet':
            res = self.stemmer.lemmatize(word)
        return res

def vectorize_text(
    corpus=None, 
    output='sparse',
    vect='count',
    param_dict=dict(
        ngram_range=(1,1),
        stop_words='english',
        max_features=1000,
        binary=False
    )
):
    """
    Helper function to vectorize raw comments in introduction.ipynb.
    Fits and transforms either a CountVectorizer ('count') or
    TfidfVectorizer ('tfidf') to a corpus of comments and returns
    a DataFrame with appropriately-labeled axes.    
    """
    if not isinstance(corpus, _pandas.Series):
        raise ValueError('Enter a Pandas Series object for corpus')
    
    if vect == 'count':
        vectorizer = _sklearn.feature_extraction.text.CountVectorizer(**param_dict)
    elif vect == 'tfidf':
        vectorizer = _sklearn.feature_extraction.text.TfidfVectorizer(**param_dict)
    else:
        raise ValueError('Unrecognized vectorizer')
    
    # Fit-transform vectorizer and put into DataFrame
    print('Fitting data...')
    vectorizer.fit(corpus)
    
    print('Transforming data...')
    sparse = vectorizer.transform(corpus)
    if output.lower() == 'sparse':
        result = sparse
    elif output.lower() == 'array': 
        result = sparse.toarray()
    elif output.lower() == 'df':
        result = _pandas.DataFrame(
            sparse.toarray(), 
            columns=vectorizer.get_feature_names(), 
            index=corpus.index)

        # Sort DF columns
        print('Sorting df cols...')
        word_frequencies = result.sum(axis=0)
        keep_cols = sorted(
            vectorizer.vocabulary_.copy(), 
            key=lambda k: word_frequencies[k], 
            reverse=True)
        result = result[keep_cols]
    
    print('Done')
    return result

toxic/common/old_common.py
import wordcloud
import matplotlib.pyplot as plt
import seaborn
from google.cloud import storage
from io import StringIO
import pandas
import numpy
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

from sklearn.base import BaseEstimator, TransformerMixin

from nltk.stem.wordnet import WordNetLemmatizer
from nltk.stem.snowball import EnglishStemmer
from nltk.stem.lancaster import LancasterStemmer

from sklearn.linear_model import LogisticRegression

import scipy

from collections import OrderedDict

class _Stemmer():
            
            def __init__(self, stemming=None):
                
                self.stemming = stemming
                
                if stemming == 'snowball':
                    self.stemmer = EnglishStemmer()
                elif stemming == 'lancaster':
                    self.stemmer = LancasterStemmer()
                elif stemming == 'wordnet':
                    self.stemmer = WordNetLemmatizer()
                else:
                    raise ValueError('Unrecognized stemmer')
            
            def stem(self, word=None):
                res = ''
                if self.stemming in ['snowball','lancaster']:
                    res = self.stemmer.stem(word)
                elif self.stemming == 'wordnet':
                    res = self.stemmer.lemmatize(word)
                return res
            
class LogisticRegression_mod(LogisticRegression):
    def __init__(self):
        return None
    
    def fit(self, x=None, y=None):
        return LogisticRegression.fit(x['matrix_res'])

class Stemmer(BaseEstimator, TransformerMixin):
    """
    Takes a term-document matrix, stems each word,
    and adds the term vectors with the same stem.
    Returns the modified vocabulary and t-d  matrix.
    """
    def __init__(self, stemming=None):
        self.stemming = stemming
        return None
    
    def fit(self, x=None, y=None, td_matrix=None, vocab=None):
        return self
    
    def transform(self, **kwargs):
        """
        Stemmer should be [snowball|lancaster|wordnet],
        see http://www.nltk.org/api/nltk.stem.html#module-nltk.stem
        for more information.
        
        pseudo-code:
        start a dict of indices to combine and the associated stem
        start a new vocabulary of stems
        start the vocab index at 0
        for each word in the vocabulary:
            stem the word
            add/append the stem's index to the result indices dict, with stem as key
            if stem not in vocab results: add at vocab index, increment vocab_index
        create a result matrix with dim(axis1) = len(vocab result dict)
        for each stem in the indices dict:
            compute np.sum(x, axis=1) where x is the columns of td_matrix at the indices in the item
            set the column of the result matrix per the result vocab dict
        """
        #print(stemming)
        print('Creating variables...')
        combine_idx = dict()
        vocab_res = OrderedDict()
        vocab_idx = 0
        stemmer = _Stemmer(stemming=self.stemming)
        
        print('Stemming vocabulary...')
        for oldword in kwargs['vocab']:
            stem = stemmer.stem(word=oldword)
            if stem in combine_idx:
                combine_idx[stem].append(kwargs['vocab'][oldword])
            else:
                combine_idx[stem] = [kwargs['vocab'][oldword]]
            
            if stem not in vocab_res:
                vocab_res[stem] = vocab_idx
                vocab_idx += 1
#                 print(vocab_res)
        
        print('Combining stem facet vectors...')
        matrix_res = scipy.sparse.csr_matrix((kwargs['td_matrix'].shape[0], 0))
        print('shape of empty matrix res:', matrix_res.shape)
        
        print('Summing across stem facets...')
#         print('matrix res', matrix_res)
        for stem in vocab_res:
            old_vectors = kwargs['td_matrix'][:, combine_idx[stem]]
#             print('old vectors shape', old_vectors.shape)
            new_vector = old_vectors.sum(axis=1)
#             print('new vector', new_vector.shape)
#             print('stem_vector:', stem_vector.sum())
#             print(matrix_res)
            matrix_res = scipy.sparse.hstack([matrix_res, new_vector], format='csr')
        
        print('result stack shape:', matrix_res.shape)
        
        print('Sorting result matrix...')
        stem_sums = matrix_res.sum(axis=0)
#         print('stem sums:', stem_sums)

        sorted_idx = numpy.asarray(numpy.argsort(stem_sums, )).flatten(order='F')[::-1]
        #
        print('sorted index shape:', sorted_idx.shape)
        print('sorted index:', sorted_idx)
        matrix_res = matrix_res[:, sorted_idx]
        
        print('Sorting vocabulary')
        #import pdb; pdb.set_trace()
        vocab_res = numpy.array([item[0] for item in vocab_res.items()])[sorted_idx]
        #def dict_to_list(dict=None):
         #   {stem: vocab_res.index(stem) for stem in vocab_res}
        #vocab_res = dict_to_list(vocab_res)
            
        return {'vocab_res': vocab_res, 'matrix_res': matrix_res}

def gcs_to_df(project='mlmasters', bucket='toxic-comments', blob='sample_submission.csv'):
    """
    Gets data from Google Cloud Storage.
    """
    client = storage.Client(project=project)
    bucket = client.get_bucket(bucket)
    s = bucket.get_blob(blob).download_as_string()
    f = StringIO(s.decode())
    df = pandas.read_csv(f)
    return df

def plot_wordcloud_to_axis(cloud_string=None, title=None, ax=None, cloud_obj=wordcloud.WordCloud(random_state=42)):
    """
    Plots a wordcloud to the passed axis using the passed string.
    """
    if cloud_string:
        cloud = cloud_obj.generate(cloud_string)
        ax.imshow(cloud)
        if title:
            ax.set_title(title)
        ax.axis("off")
        return ax

def density_plot(data=None, variable=None, color_mask=None):
    """
    somewhat ggplot-like simple plotting 
    interface for densities segmented by a mask
    """
    fig = plt.figure(1, figsize=(10,5))
    
    ax1 = plt.subplot(121)
    barvals = data.groupby(color_mask).mean()[variable]
    seaborn.barplot(x=barvals.index, y=barvals, ax=ax1)
    
    ax2 = plt.subplot(122)
    seaborn.kdeplot(data.loc[~color_mask, variable], shade=True, color="b", label='normal', ax=ax2)
    seaborn.kdeplot(data.loc[color_mask, variable], shade=True, color="g", label='toxic', ax=ax2)
    return fig

default_params = dict(
    ngram_range=(1,1),
    stop_words='english',
    max_features=1000,
    binary=False
)

def vectorize_text(
    corpus=None, 
    output='sparse',
    vect='count',
    param_dict=default_params
):
    """
    Helper function to vectorize raw comments.
    Fits and transforms either a CountVectorizer ('count') or
    TfidfVectorizer ('tfidf') to a corpus of comments and returns
    a DataFrame with appropriately-labeled axes.    
    """
    if not isinstance(corpus, pandas.Series):
        raise ValueError('Enter a Pandas Series object for corpus')
    
    if vect == 'count':
        vectorizer = CountVectorizer(**param_dict)
    elif vect == 'tfidf':
        vectorizer = TfidfVectorizer(**param_dict)
    else:
        raise ValueError('Unrecognized vectorizer')
    
    # Fit-transform vectorizer and put into DataFrame
    print('Fitting data...')
    vectorizer.fit(corpus)
    
    print('Transforming data...')
    sparse = vectorizer.transform(corpus)
    if output.lower() == 'sparse':
        result = sparse
    elif output.lower() == 'array': 
        result = sparse.toarray()
    elif output.lower() == 'df':
        result = pandas.DataFrame(
            sparse.toarray(), 
            columns=vectorizer.get_feature_names(), 
            index=corpus.index)

        # Sort DF columns
        print('Sorting df cols...')
        word_frequencies = result.sum(axis=0)
        keep_cols = sorted(
            vectorizer.vocabulary_.copy(), 
            key=lambda k: word_frequencies[k], 
            reverse=True)
        result = result[keep_cols]
    
    print('Done')
    return result